### DISCUSSION TITLE
Wagmi redirects me to App Store IOS

### DISCUSSION THREAD

**Discussion by arash-hk** [wevm/wagmi#2115] [Q&A] [answered]
hi everyone, 
i have a problem with wallet on my **IOS** device and when i want to write to contract when metamask is connected, it redirects me to app store  !

![IMG_20230403_021139](https://user-images.githubusercontent.com/111232556/229383068-85b4cdb5-d4d9-4527-bf9b-06ab34ae4090.jpg)

i am using nextjs, wagmi and rainbowkit !

and here is my code:

**_1. app.js _**

```
import "@rainbow-me/rainbowkit/styles.css";
import {
  connectorsForWallets,
  RainbowKitProvider,
  darkTheme,
} from "@rainbow-me/rainbowkit";
import { configureChains, createClient, WagmiConfig } from "wagmi";
import {
  mainnet,
  goerli,
  polygon,
  sepolia,
  polygonMumbai,
  optimism,
  optimismGoerli,
} from "wagmi/chains";
import { publicProvider } from "wagmi/providers/public";
import {
  rainbowWallet,
  walletConnectWallet,
  metaMaskWallet,
  omniWallet,
  trustWallet,
  imTokenWallet,
  argentWallet,
  coinbaseWallet,
  ledgerWallet,
} from "@rainbow-me/rainbowkit/wallets";

const { chains, provider, webSocketProvider } = configureChains(
  [mainnet, goerli, polygon, sepolia, polygonMumbai, optimism, optimismGoerli],
  [publicProvider()]
);

const connectors = connectorsForWallets([
  {
    groupName: "Recommended",
    wallets: [
      metaMaskWallet({ chains }),
      walletConnectWallet({ chains }),
      rainbowWallet({ chains }),
      coinbaseWallet({ chains }),
    ],
  },
  {
    groupName: "Others",
    wallets: [
      trustWallet({ chains }),
      ledgerWallet({ chains }),
      argentWallet({ chains }),
      omniWallet({ chains }),
      imTokenWallet({ chains }),
    ],
  },
]);

const wagmiClient = createClient({
  autoConnect: true,
  connectors,
  provider,
});

export default function App({ Component, pageProps }) {
  return (
    <WagmiConfig client={wagmiClient}>
      <RainbowKitProvider
        modalSize="compact" //wide,compact
        chains={chains}
        theme={darkTheme({
          accentColor: "black",
          accentColorForeground: "white",
          borderRadius: "medium",
          fontStack: "system",
          overlayBlur: "small",
        })}
      >
        <Component {...pageProps} />
      </RainbowKitProvider>
    </WagmiConfig>
  );
}

```

**_2. index.js_**

```
import Head from "next/head";
import Image from "next/image";
import { Inter } from "@next/font/google";
import styles from "@/styles/Home.module.css";
const inter = Inter({ subsets: ["latin"] });
import { ConnectButton } from "@rainbow-me/rainbowkit";
import MintPage from "@/MintPage/MintPage";

export default function Home() {
  return (
    <>
      <ConnectButton />
      <MintPage />
    </>
  );
}
```

**_3.package.json_**

```
{
  "name": "my-app",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@next/font": "13.1.6",
    "@rainbow-me/rainbowkit": "^0.11.0",
    "ethers": "^5",
    "next": "13.1.6",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "wagmi": "^0.11.6"
  }
}
```

if you want to checkout mintnow that was in index.js here it is:
```

import {
  useContractRead,
  usePrepareContractWrite,
  useContractWrite,
} from "wagmi";
import contractABI from "../constants/abi.json";
import { ethers } from "ethers";
const contractAddress = "0x00664DC7C5C79A666a3Ed176D1Cd7f2cbd4165E1";

export default function MintPage() {

  //USE PREPARE CONTRACT WRITE

  //Public_Mint_Common
  
  const { config: public_mint_common_config, refetch: refetch_common } = usePrepareContractWrite({
    address: contractAddress,
    abi: contractABI,
    functionName: "Public_Mint_Common",
    args: [1],
    enabled: false,
    overrides: {
      value: ethers.utils.parseEther("0.1"),
    },
  });

  //USE CONTRACT WRITE
  
  //public_mint_common

  const { write:public_mint_common } = useContractWrite(public_mint_common_config);
  
  const mint_common = async () => {
    await refetch_common()
    public_mint_common?.()
  }

  return (
    <>
      <div>
        <button onClick={ ()=> mint_common() }>public mint common</button>
      </div>
    </>
  );
}


```
**Current Behavior:**

redirect from metamask icon in rainbowkit to app store!

**Expected Behavior:**

work correctly like android devices and laptop users.

**can you help me please?!**

**Comment by jxom** [comment, marked as answer]
You shouldn’t be refetching in `mint_common` as this causes a long-running asynchronous task which leads to the App Store prompt you are seeing: https://wagmi.sh/react/prepare-hooks#ios-app-link-constraints

The Prepare Hook shouldn’t have `enabled: false` or the  refetch behavior within `mint_common`. It should work without.

**Reply by arash-hk** [reply]
thanks for you valuable time dear jxom, but without details you said above the code will be like this:

![12](https://user-images.githubusercontent.com/111232556/229475786-92da0f7e-ec6c-4beb-bdde-1463154bbbc6.PNG)


but there is another problem! i opened a discussion a few days ago and shared this problem, when the code is like the picture, it automaticly get called when i refresh my react app on chrome!
here is the discussion link :https://github.com/wagmi-dev/wagmi/discussions/2037

look at this:

![13](https://user-images.githubusercontent.com/111232556/229475528-74c5bfda-6c66-48f2-9fa1-d9a3c1f2ac49.PNG)

i want to run the write function onClick! what should i do now?

**Reply by jxom** [reply]
The write function isn’t being called when the page loads. Only the validation gets run. In the above screenshot, you have no funds, so the Hook is eagerly letting you know that you won’t be able to execute the write function. 

See more: https://wagmi.sh/react/prepare-hooks#no-upfront-validation

You may want to disable the mint button and display an error to the user that they don’t have enough funds to mint.

**Reply by arash-hk** [reply]
Thank you again and wish you all the best!

You solved an annoying question of me and lots of developers

I mark your reply as answer

**Comment by petermunachi** [comment]
Hello guys i face this issue when i try to connect to metamask on Iphone 12 pro. I get redirected to app store even though i already have Metamask installed

**Reply by tmm** [reply]
This is a WalletConnect limitation that will be fixed in the future when WalletConnect moves to deeplinks instead of universal links.

In the meantime, here's how you can fix it: https://twitter.com/_DanielSinclair/status/1622751806789910528
